packaging/metadata.py

```diff
@@ -0,0 +1,387 @@
+# This file is dual licensed under the terms of the Apache License, Version
+# 2.0, and the BSD License. See the LICENSE file in the root of this repository
+# for complete details.
+
+import dataclasses
+import math
+import sys
+from email import message_from_bytes
+from email.headerregistry import Address, AddressHeader
+from email.message import EmailMessage
+from email.policy import EmailPolicy, Policy
+from functools import reduce
+from inspect import cleandoc
+from itertools import chain
+from typing import (
+    TYPE_CHECKING,
+    Any,
+    ClassVar,
+    Collection,
+    Dict,
+    Iterable,
+    Iterator,
+    Mapping,
+    Set,
+    Tuple,
+    Type,
+    TypeVar,
+    Union,
+    cast,
+)
+
+from .requirements import InvalidRequirement, Requirement
+from .specifiers import SpecifierSet
+from .version import Version
+
+T = TypeVar("T", bound="CoreMetadata")
+A = TypeVar("A")
+B = TypeVar("B")
+
+if sys.version_info[:2] >= (3, 8) and TYPE_CHECKING:  # pragma: no cover
+    from typing import Literal
+
+    NormalizedDynamicFields = Literal[
+        "platform",
+        "summary",
+        "description",
+        "keywords",
+        "home-page",
+        "author",
+        "author-email",
+        "license",
+        "supported-platform",
+        "download-url",
+        "classifier",
+        "maintainer",
+        "maintainer-email",
+        "requires-dist",
+        "requires-python",
+        "requires-external",
+        "project-url",
+        "provides-extra",
+        "provides-dist",
+        "obsoletes-dist",
+        "description-content-type",
+    ]
+else:
+    NormalizedDynamicFields = str
+
+
+def _normalize_field_name_for_dynamic(field: str) -> NormalizedDynamicFields:
+    """Normalize a metadata field name that is acceptable in `dynamic`.
+
+    The field name will be normalized to lower-case. JSON field names are
+    also acceptable and will be translated accordingly.
+
+    """
+    return cast(NormalizedDynamicFields, field.lower().replace("_", "-"))
+
+
+# Bypass frozen dataclass for __post_init__, this approach is documented in:
+# https://docs.python.org/3/library/dataclasses.html#frozen-instances
+_setattr = object.__setattr__
+
+
+# In the following we use `frozen` to prevent inconsistencies, specially with `dynamic`.
+# Comparison is disabled because currently `Requirement` objects are
+# unhashable/not-comparable.
+
+
+@dataclasses.dataclass(frozen=True, eq=False)
```

Comment:
That list of arguments is massive. Maybe it would be best to require keywords only? Otherwise there's no way someone is going to know what e.g. parameter 7 is.

```suggestion
@dataclasses.dataclass(frozen=True, eq=False, kw_only=True)
```